### ISSUE TITLE
inject constraints & hints into generation process

### ISSUE THREAD

**Issue by zeeshanlakhani** [zeeshanlakhani/schema-gen#2]
I've done some work on this that's really specific to our data at [sfx](https://github.com/sfx) (and within private repos). Will look into doing something more general.
- Dive into Prismatic coercers approach

[Related Issue/Comment](https://github.com/Prismatic/schema/issues/103#issuecomment-45419161)

**Comment by dmcollie** [comment]
Are there any hints about how to handle predicates in the schema definitions? At the moment I'm getting an exception thrown but can I add a multi method for schema->gen* for a specific key somehow?

**Comment by gfredericks** [comment]
predicates are tricky because in test.check a predicate used with `such-that` has to be highly likely to pass, w.r.t. the generator it's used with. Meanwhile in schema it's totally reasonable and probably common to use a predicate that would be highly likely to fail for a natural distribution over the relevant space.

I'd like this to be better but don't have any general ideas. I feel like it would likely involve using a more specific set of schema tools that are designed with generation in mind. If you describe your current use case I could speculate about how that might fit.

**Comment by zeeshanlakhani** [comment]
@gfredericks's right. Predicates make things difficult, but it's also something I've wanted to put more thought into.

**Comment by dmcollie** [comment]
I have a fairly straightforward constraint just now - strings of a maximum length. Ideally the generated values would respect this constraint. Would an approach be to hint the generator using metadata?

Thanks


Dave.

**Comment by gfredericks** [comment]
yeah some kind of special schema that has info for schema-gen would work. You couldn't use `String` directly since a class can't have metadata, but I think both schema and schema-gen are extensible enough that this kind of thing can work. I'm just not sure if there is any kind of useful general solution, or if you'd end up fully specifying the generator every time.
